### DISCUSSION TITLE
Monte Carlo code where it divide the number of samples

### DISCUSSION THREAD

**Discussion by NhatMinh2001** [mitsuba-renderer/mitsuba3#1197] [Q&A] [answered]
Hi Mitsuba developing team and everyone else, I want to ask this silly question. As i known, a monte carlo estimator aggregates samples and the divide for the number of samples.
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/172172950/d5adc79a-1af2-4503-8a9a-03e7f5877acb)
But I've not yet found the code where it do that. **could you point me the part where the algorithm divide for `spp`?**.
**Also, i want to set different parts of image with different, adaptable `spp`  (not known before, like adaptive sampling), what are the right ways to do that?**
Big thank to all the developers and contributors who making this project happen.

**Comment by njroussel** [comment, marked as answer]
Hi @NhatMinh2001 

This is split over a few different files. An integrator will typically [return a "weight" for each sample](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/render/integrator.cpp#L485), and its value is usually 1. These weights are accumulated into their own channel in the film, like RGB values. Finally the weight is [divided out when developing the film](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/films/hdrfilm.cpp#L400). So effectively, this weight channel is equivalent to `spp`.

For your questions about varying spp, it is handled implicitly. The function `Sensor.sample_rays()` generates the rays with a constant spp, but it could do so according to a completely arbitrary distribution. So if you want to do some adaptive sampling like technique, modifiying the `sample_rays()` method should be enough. The change in distribution will automatically be reflected in the accumulated weight channel, and hence the `develop()` method will divide out the correct value -- nothing needs to be changed there.
